File: 3-1.c
#include "prelude.h"

// priorities go from [1, 52]
#define N 53
#define LEFT 1
#define RIGHT 2

static char const example[] = "vJrwpWtwJgWrhcsFMMfFFhFp\n"
			      "jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL\n"
			      "PmmdzqPrVvPwwTWBwg\n"
			      "wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn\n"
			      "ttgJtRGJQctTZtZT\n"
			      "CrZsJsPPZsGzwwsLwLmpwMDw";

static int priority(char c) {
	assert(c <= 'z' && c >= 'A' && "Unexpected character!");
	return c >= 'a' && c <= 'z' ? c - 'a' + 1 : c - 'A' + 27;
}

static int solve(char *input) {
	int items[N] = {0};
	int sum = 0;

	LINES(line, input) {
		size_t n = strlen(line);
		for (size_t i = 0; i < n / 2; i++)
			items[priority(line[i])] |= LEFT;
		for (size_t i = n / 2; i < n; i++)
			items[priority(line[i])] |= RIGHT;

		for (size_t i = 1; i < N; i++) {
			if (items[i] == (LEFT | RIGHT)) sum += i;
			items[i] = 0;
		}
	}

	return sum;
}

int main() {
	RUN(solve, strdup(example));
	RUN(solve, read_to_string("3.txt"));
	return 0;
}

File: rs/src/bin/3-1.rs
const EXAMPLE: &str = r"vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw";

/// Partition into two equal length parts
fn partition(s: &str) -> (&str, &str) {
    assert!(s.len() % 2 == 0);
    s.split_at(s.len() / 2)
}

/// Return the numeric priority of each char
fn priority(c: char) -> i32 {
    match c {
        'a'..='z' => 1 + c as i32 - 'a' as i32,
        'A'..='Z' => 27 + c as i32 - 'A' as i32,
        _ => panic!("Invalid character '{}'", c),
    }
}

/// Return the single duplicate character
fn duplicate((left, right): (&str, &str)) -> char {
    left.chars()
        .find(|&c| right.contains(c))
        .expect("There must be a duplicate!")
}

fn solve(input: &str) -> i32 {
    input
        .lines()
        .map(partition)
        .map(duplicate)
        .map(priority)
        .sum()
}

fn main() -> std::io::Result<()> {
    println!("Example: {}", solve(EXAMPLE));
    println!("Example: {}", solve(&std::fs::read_to_string("../3.txt")?));
    Ok(())
}
